Optimizations for heavy multithreaded usage

muryoh opened the pull request:
Method cache was implemented using Collections.synchronizedMap()
which is bad under heavy multi threading usage
Replaced it with ConcurrentHashMap which is way better, with
a nice hack as it doesn't allow null values :o

+ optimized Dependency.onGet()

muryoh committed:
Use ConcurrentHashMap in InstanceManager

Method cache was implemented using Collections.synchronizedMap()
which is bad under heavy multi threading usage
Replaced it with ConcurrentHashMap which is way better, with
a nice hack as it doesn't allow null values :o

Change-Id: Iee4fff134fa59dee68a7db9c78af92938d0ebe10

muryoh committed:
Optimize Depency.onGet()

This method can get called a lot depending on the caller usages
When it is, the previous implementation would basically recompute
the best object to be returned, locking and unlocking on its way to
make sure consistency was guaranteed, and clearing the returned
object reference after use
Now the returned object reference is actually re-used as long as no
service modification occurs. As soon as one does occur, the object
reference is marked as "to be re-computed"

VincentVk9373 commented on core/src/main/java/org/apache/felix/ipojo/handlers/dependency/Dependency.java:
```diff
@@ -661,13 +673,15 @@ public Object getService() {
         Usage usage = (Usage) m_usage.get();
```
Here there is no null check on m_usage... it is a bit sad to have an inconsistency...

Can it really be null on lines 595 & 611 ?

muryoh replied:
Honestly, hard to say

In some cases it is checked, like in onEntry()/onExit()
Since the initialization is done with
```
        if (field != null) {
            m_usage = new ServiceUsage();
        } else {
            m_usage = null;
        }```
I'm guessing it has to do with the code paths depending on if a field was provided or not
I didn't have enough knowledge and didn't want to risk it so I just did the check :)

muryoh committed:
Fix InstanceManager.getMethodById()

It wasn't actually caching the computed results for
inner class methods as it was not using the correct methodId key
